### PR TITLE
Fix type hints in typing edu unit tests

### PR DESCRIPTION
**Pull request by anoadragon453**
Best to review commit-by-commit. Commits contain individual descriptions of changes.

### CODE REVIEW

File: synapse/server.py [hunk 1]
```diff
@@ -479,7 +489,7 @@ def get_presence_router(self) -> PresenceRouter:
         return PresenceRouter(self)
 
     @cache_in_self
-    def get_typing_handler(self) -> FollowerTypingHandler:
+    def get_typing_handler(self) -> Union[TypingWriterHandler, FollowerTypingHandler]:
```
Comment (clokep) [inline synapse/server.py:492]: I don't think we need this change, `TypingWriterHandler` is a sub-class of `FollowerTypingHandler`.

File: tests/handlers/test_typing.py [hunk 2]
```diff
@@ -98,7 +99,13 @@ def prepare(self, reactor: MemoryReactor, clock: Clock, hs: HomeServer) -> None:
...
         self.on_new_event = mock_notifier.on_new_event
 
-        self.handler = hs.get_typing_handler()
+        self.handler = cast(TypingWriterHandler, hs.get_typing_handler())
+
+        # hs.get_typing_handler will return a TypingWriterHandler when calling it
+        # from the main process, and a FollowerTypingHandler on workers.
+        # We rely on methods only available on the former, so assert we have the
+        # correct type here.
+        self.assertIsInstance(self.handler, TypingWriterHandler)
```
Comment (clokep) [inline tests/handlers/test_typing.py:107]: We should just be able to do the `assert` without the `cast`; the trick is you need to use a bare `assert` or mypy won't realize it is being used for type narrowing.

```python
handler = hs.get_typing_handler()
assert isinstance(handler, TypingWriterHandler)
self.handler = handler

```

Comment (anoadragon453) [reply tests/handlers/test_typing.py]: Nice trick. So mypy needs to have the assert before you assign the type of the variable for the first time?

Comment (clokep) [reply tests/handlers/test_typing.py]: Only because you're using it in other scopes. Otherwise it thinks the property in the class is of the wrong type and you need to assert each time you use it.

File: tests/handlers/test_typing.py [hunk 3]
```diff
@@ -62,7 +63,11 @@ def _make_edu_transaction_json(edu_type: str, content: JsonDict) -> bytes:
 
 
 class TypingNotificationsTestCase(unittest.HomeserverTestCase):
-    def make_homeserver(self, reactor: MemoryReactor, clock: Clock) -> HomeServer:
+    def make_homeserver(
+        self,
+        reactor: ThreadedMemoryReactorClock,
```
Comment (clokep) [inline tests/handlers/test_typing.py:68]: Why the change from `MemoryReactor`? It isn't clear to me what's wrong with that?

Comment (clokep) [reply tests/handlers/test_typing.py:68]: This seems to be because we use `pump()` which is defined on `MemoryReactorClock`. And turns out that the super-class *does* use `ThreadedMemoryReactorClock`.

**Review by clokep — APPROVED** [review]

File: tests/handlers/test_typing.py [hunk 4]
```diff
@@ -14,21 +14,22 @@
 
 
 import json
-from typing import Dict
+from typing import Dict, List, Set, cast
```
Comment (clokep) [inline tests/handlers/test_typing.py:17]:
```suggestion
from typing import Dict, List, Set
```

**Comment by anoadragon453** [comment]
CI failures are unrelated to this PR (flake on `TestPartialStateJoin` Complement test).
